src/main/java/thaumicenergistics/common/tiles/TileAdvancedInfusionProvider.java

```diff
@@ -0,0 +1,235 @@
+            for (Aspect aspect : aspectList.getAspects()) {
+                if (aspect != null) {
+                    int needAspectsAmount = aspectList.getAmount(aspect);
+                    int extractAmount = this.extractEssentiaFromNetwork(aspect, needAspectsAmount, false);
+
+                    if (needAspectsAmount != extractAmount) {
+                        this.orderSomeEssentia(aspect, needAspectsAmount - extractAmount);
+                    }
+                    this.doParticalFX(aspect.getColor());
+                    matrix.getAspects().remove(aspect, extractAmount);
+                }
+            }
+        }
+    }
+
+    @Override
+    @TileEvent(TileEventType.WORLD_NBT_WRITE)
+    public void onSaveNBT(final NBTTagCompound data) {
+        if (this.matrix != null) {
+            data.setInteger(NBT_MATRIX_X, this.matrix.xCoord);
+            data.setInteger(NBT_MATRIX_Y, this.matrix.yCoord);
+            data.setInteger(NBT_MATRIX_Z, this.matrix.zCoord);
+        }
+        data.setBoolean(NBT_HAS_ADD_STABILITY, this.hasAddStability);
+        super.onSaveNBT(data);
+    }
+
+    @Override
+    @TileEvent(TileEventType.WORLD_NBT_READ)
+    public void onLoadNBT(final NBTTagCompound data) {
+        if (data.hasKey(NBT_MATRIX_X) && data.hasKey(NBT_MATRIX_Y) && data.hasKey(NBT_MATRIX_Z)) {
+            this.matrixX = data.getInteger(NBT_MATRIX_X);
+            this.matrixY = data.getInteger(NBT_MATRIX_Y);
+            this.matrixZ = data.getInteger(NBT_MATRIX_Z);
+        }
+        if (data.hasKey(NBT_HAS_ADD_STABILITY)) {
+            this.hasAddStability = data.getBoolean(NBT_HAS_ADD_STABILITY);
+        }
+        super.onLoadNBT(data);
+    }
+
+    @Override
+    public void addWailaInformation(List<String> tooltip) {
+        super.addWailaInformation(tooltip);
+        if (this.matrix == null) {
+            tooltip.add(
+                    ThEStrings.Tooltip_AdvancedInfusionProviderWorkingMode.getLocalized() + ":"
+                            + ThEStrings.Tooltip_AdvancedInfusionProviderNormalMode.getLocalized());
+        } else {
+            tooltip.add(
+                    ThEStrings.Tooltip_AdvancedInfusionProviderWorkingMode.getLocalized() + ":"
+                            + ThEStrings.Tooltip_AdvancedInfusionProviderAdvancedMode.getLocalized());
+            tooltip.add(
+                    String.format(
+                            ThEStrings.Tooltip_AdvancedInfusionProviderBindTo.getLocalized(),
+                            this.matrixX,
+                            this.matrixY,
+                            this.matrixZ));
+        }
+    }
+
+    @TileEvent(TileEventType.TICK)
+    public void onTick() {
+        if (this.matrix == null && this.isActive) {
+            if (this.matrixX == null || this.matrixY == null || this.matrixZ == null) {
+                this.searchMatrix();
+            } else {
+                this.bindMatrix();
+            }
```

Comment:
Search for matrix in 9x9x9 area should not be performed every tick. Add a counter variable and increment it every tick. And search for matrix every 10th or 20th tick.
```java
if (++tickCounter % 20 == 0) {
	if (this.matrixX == null || this.matrixY == null || this.matrixZ == null) {
        this.searchMatrix();
    }
}
```
You're also don't need to rebind matrix every tick - it is bound once during search procedure.